File: code-monkey/src/agents/agent.py
import os
from typing import Set, Any
from langgraph.prebuilt import create_react_agent
from langchain_core.messages import HumanMessage, SystemMessage

from tools.utils import (
    ask_user,
    show_diff,
)
from constants import get_root_dir, get_artifacts_dir, get_agent_msn
from models import MSN
from .base_agent import BaseAgent
from instrumentation import current_span, instrument
from langgraph.checkpoint.aiosqlite import AsyncSqliteSaver

from util.logs import get_logger

class Agent(BaseAgent):
    # trying to figure out how to type this.  It's not a BaseChatModel, it's a
    # CompiledGraph from langgraph.graph, but that's not exported
    model: Any

    config = {
        "configurable": {"thread_id": "abc123"},
    }

    @instrument("Agent.__init__", ["msn_str"])
    def __init__(self):
        msn_str = get_agent_msn()
        msn = MSN.from_string(msn_str)

        # a checkpointer + the thread_id below gives the model a way to save its
        # state so we don't have to accumulate the messages ourselves.
        memory = AsyncSqliteSaver.from_conn_string(":memory:")
        self.model = create_react_agent(
            msn.construct_model(),
            self.tools,
            checkpointer=memory,
        )
        self.initialize()

    # Custom Agent initialization goes here, when necessary.
    def initialize(self):
        pass

    def prepare_prompt(self, prompt: str) -> str:
        return prompt

    @instrument("Agent.run_prompt", ["prompt"])
    async def run_prompt(self, prompt: str) -> str:
        logger = get_logger(__name__)
        logger.info(f"Running prompt: {prompt}")

        system = SystemMessage(content=self.get_system_prompt())
        msg = HumanMessage(content=self.prepare_prompt(prompt))

        modified_files: Set[str] = set()
        logger = get_logger(__name__)
        
        result: str | None = None

        async for event in self.model.astream_events(
            {
                "messages": [system, msg],
            },
            self.config,
            version="v2",
        ):
            kind = event["event"]

            if kind != "on_chat_model_stream" and kind != "on_chain_end" and kind != "on_chain_stream" and kind != "on_chain_start":
                logger.debug(f"[AGENT {self.name}] received event: {kind}")

            if kind == "on_chat_model_start":
                # for msg in event["data"]["input"]["messages"][0]:
                #     print("----")
                #     print(msg.content)
                continue

            if kind == "on_chat_model_stream":
                # NB(toshok) we can stream text as it comes in by doing it here,
                # but I'm doing the entire thing in one go in the
                # on_chat_model_end event just below.
                continue

            if kind == "on_tool_start":
                tool_name = event["name"]
                tool_input = event["data"]["input"]
                logger.debug(f"tool start - name={tool_name}, input={repr(tool_input)}")
                continue

            if kind == "on_tool_end":
                logger.debug("----")
                tool_name = event["name"]
                tool_output = event["data"]["output"].content[:20]
                logger.debug(f"[AGENT {self.name}] tool end - name={tool_name}, output={repr(tool_output)}")
                continue

            if kind == "on_chat_model_end":
                if isinstance(event["data"]["output"].content, list):
                    # anthropic seems give us content as a list?
                    result = event["data"]["output"].content[0]["text"]
                else:
                    result = event["data"]["output"].content
                print(f"[AGENT {self.name}]" + result)
                # TODO(toshok) still need to compute tokens
                continue

            if kind == "on_custom_event":
                # TODO(toshok) a better dispatch mechanism would be nice, but
                # there's only one event type currently
                if event["name"] == "file_modified":
                    modified_files.add(event["data"])
                else:
                    logger.debug(f"[AGENT {self.name}] Unknown prompt event: '{event["name"]}'")
                continue

        self.handle_completion(
            modified_files,
        )

        # TODO(toshok) we aren't returning a result here, and likely should...
        logger.debug(f"[AGENT {self.name}] run_prompt result: \"{result or ""}\"")
        return result or ""

    @instrument("Agent.handle_completion")
    def handle_completion(self, modified_files: set) -> None:
        current_span().set_attributes(
            {
                "num_modified_files": len(modified_files),
                "modified_files": str(modified_files),
            }
        )

        if modified_files:
            for file in modified_files:
                self.handle_modified_file(file)

    @instrument("Agent.handle_modified_file", ["file"])
    def handle_modified_file(self, file: str) -> None:
        original_file = os.path.join(get_root_dir(), file)
        modified_file = os.path.join(get_artifacts_dir(), file)
        show_diff(original_file, modified_file)
        response = ask_user(
            f"Do you want to apply the changes to {file} (diff shown in VSCode)? (Y/n): "
        ).lower()

        apply_changes = response == "y" or response == ""

        current_span().set_attribute("apply_changes", apply_changes)

        if apply_changes:
            self.apply_changes(original_file, modified_file)
            print(f"✅ Changes applied to {file}")
        else:
            print(f"❌ Changes not applied to {file}")

    @instrument("Agent.apply_changes", ["original_file", "modified_file"])
    def apply_changes(self, original_file: str, modified_file: str) -> None:
        # TODO(toshok) we probably want the before/after size?  or something about size of the diff
        with open(modified_file, "r") as modified, open(original_file, "w") as original:
            original.write(modified.read())

File: code-monkey-ts2/agents/agent.ts
import fs from "fs";
import path from "path";
import { randomUUID } from "crypto";

import { createReactAgent } from "@langchain/langgraph/prebuilt";
import { MemorySaver } from "@langchain/langgraph";
import { HumanMessage, SystemMessage } from "@langchain/core/messages";
import { StructuredTool } from "@langchain/core/tools";

import { instrument, currentSpan } from "../instrumentation";
import { getAgentConfig } from "../config";
import { getRootDir, getArtifactsDir } from "../constants";
import { MSN } from "../models";
import { showDiff, askUser } from "../tools/utils";

import { BaseAgent } from "./base_agent";

function debugLog(...args: any[]) {
  console.debug(...args);
}

export abstract class Agent extends BaseAgent {
  model: ReturnType<typeof createReactAgent>;

  config = {
    configurable: { thread_id: randomUUID() },
  };

  // TODO(class decorator here) @instrument("Agent.__init__", ["msn_str"])
  constructor(name: string, systemPrompt: string, tools: StructuredTool[]) {
    super(name, systemPrompt, tools);

    console.log(`Agent ${name} constructor`);
    const { msn: msn_str } = getAgentConfig(name);
    const msn = MSN.from_string(msn_str);

    // a checkpointer + the thread_id below gives the model a way to save its
    // state so we don't have to accumulate the messages ourselves.
    const checkpointer = new MemorySaver();
    this.model = createReactAgent({
      llm: msn.constructModel(),
      tools: this.tools,
      checkpointSaver: checkpointer,
    });
  }

  async initialize(): Promise<void> {}

  preparePrompt(prompt: string): string {
    return prompt;
  }

  @instrument("Agent.runPrompt")
  async runPrompt(prompt: string): Promise<string> {
    currentSpan().setAttributes({
      agent: this.name,
      prompt,
    });
    // TODO
    // logger = get_logger(__name__)
    // logger.info(`Running prompt: {prompt}`);

    console.log(`[AGENT ${this.name}] Running prompt: ${prompt}`);
    const system = new SystemMessage({
      content: this.systemPrompt,
    });
    const msg = new HumanMessage({
      content: this.preparePrompt(prompt),
    });

    const modified_files: Set<string> = new Set();
    let result: string | null = null;

    for await (const event of this.model.streamEvents(
      {
        messages: [system, msg],
      },
      {
        configurable: this.config.configurable,
        version: "v2",
      },
    )) {
      const kind = event.event;

      if (
        kind != "on_chat_model_stream" &&
        kind != "on_chain_end" &&
        kind != "on_chain_stream" &&
        kind != "on_chain_start"
      ) {
        debugLog(
          `[AGENT ${this.name}] received event: ${JSON.stringify(event, null, 2)}`,
        );
      }

      if (kind == "on_chat_model_start") {
        // for msg in event.data.input["messages"][0]:
        //     print("----")
        //     print(msg.content)
        continue;
      }

      if (kind == "on_chat_model_stream") {
        // NB(toshok) we can stream text as it comes in by doing it here,
        // but I'm doing the entire thing in one go in the
        // on_chat_model_end event just below.
        continue;
      }

      if (kind == "on_tool_start") {
        // TODO
        const tool_name = event.name;
        const tool_input = event.data.input;
        debugLog(
          `[AGENT ${this.name}] ${kind} - name=${tool_name}, input=${JSON.stringify(tool_input)}`,
        );
        continue;
      }

      if (kind == "on_tool_end") {
        const tool_name = event.name;
        const tool_output = event.data.output.content.slice(0, 20);
        debugLog(
          `[AGENT ${this.name}] ${kind} - name=${tool_name}, output=${JSON.stringify(tool_output)}`,
        );
        continue;
      }

      if (kind == "on_chat_model_end") {
        if (Array.isArray(event.data.output["content"])) {
          // anthropic seems give us content as a list?
          result = event.data.output["content"][0]["text"];
        } else {
          result = event.data.output.content;
        }
        debugLog(`[AGENT ${this.name}] RESULT: ${result}`);
        // TODO(toshok) still need to compute tokens
        continue;
      }

      if (kind == "on_custom_event") {
        // TODO(toshok) a better dispatch mechanism would be nice, but
        // there's only one event type currently
        if (event.name === "file_modified") {
          modified_files.add(event.data as string);
        } else {
          debugLog(
            `[AGENT ${this.name}] Unknown prompt event: '${event["name"]}'`,
          );
        }
        continue;
      }
    }

    await this.handleCompletion(modified_files);

    debugLog(`[AGENT ${this.name}] runPrompt result: "${result || ""}"`);
    return result || "";
  }

  @instrument("Agent.handleCompletion")
  async handleCompletion(modified_files: Set<string>): Promise<void> {
    currentSpan().setAttributes({
      agent: this.name,
      num_modified_files: modified_files.size,
      modified_files: Array.from(modified_files).join(", "),
    });

    for (const file of Array.from(modified_files)) {
      await this.handleModifiedFile(file);
    }
  }

  @instrument("Agent.handle_modified_file")
  async handleModifiedFile(file: string): Promise<void> {
    currentSpan().setAttributes({
      agent: this.name,
      file,
    });
    const original_file = path.join(getRootDir(), file);
    const modified_file = path.join(getArtifactsDir(), file);
    showDiff(original_file, modified_file);
    const response = (
      await askUser(
        `Do you want to apply the changes to ${file} (diff shown in VSCode)? (Y/n): `,
      )
    ).toLowerCase();

    const apply_changes = response == "y" || response == "";

    // TODO
    // current_span().set_attribute("apply_changes", apply_changes)

    if (apply_changes) {
      this.applyChanges(original_file, modified_file);
      console.log(`✅ Changes applied to ${file}`);
    } else {
      console.log(`❌ Changes not applied to ${file}`);
    }
  }

  @instrument("Agent.applyChanges")
  applyChanges(original_file: string, modified_file: string): void {
    currentSpan().setAttributes({
      agent: this.name,
      original_file,
      modified_file,
    });

    // TODO(toshok) we probably want the before/after size?  or something about size of the diff
    const modified_content = fs.readFileSync(modified_file, "utf8");
    fs.writeFileSync(original_file, modified_content);
  }
}
